hadoop/device-plugins/gpu-mig/src/main/java/com/nvidia/spark/NvidiaGPUMigPluginForRuntimeV2.java

```diff
@@ -0,0 +1,366 @@
+            output = shellExecutor.getMajorMinorInfo(devName);
+            String[] strs = output.trim().split(":");
+            output = Integer.toString(Integer.parseInt(strs[0], 16));
+        } catch (IOException e) {
+            String msg =
+                    "Failed to get major number from reading /dev/" + devName;
+            LOG.warn(msg);
+        } catch (NumberFormatException e) {
+            LOG.error("Failed to parse device major number from stat output");
+            output = null;
+        }
+        return output;
+    }
+
+    @Override
+    public Set<Device> allocateDevices(Set<Device> availableDevices, int count,
+                                       Map<String, String> envs) {
+        Set<Device> allocation = new TreeSet<>();
+        String envShouldThrow = envs.get(THROW_MULTI_ENV);
+        if (envShouldThrow != null) {
+            shouldThrowOnMultipleGPUFromEnv = Boolean.parseBoolean(envShouldThrow);
+        } else {
+            shouldThrowOnMultipleGPUFromEnv = true;
+        }
+        // Only officially support 1 GPU per container so don't worry about topology
+        // scheduling.
+        basicSchedule(allocation, count, availableDevices);
+        return allocation;
+    }
+
+    public void basicSchedule(Set<Device> allocation, int count,
+                              Set<Device> availableDevices) {
+        // Basic scheduling
+        // allocate all available
+        if (count == availableDevices.size()) {
+            allocation.addAll(availableDevices);
+            return;
+        }
+        int number = 0;
+        for (Device d : availableDevices) {
+            allocation.add(d);
+            number++;
+            if (number == count) {
+                break;
+            }
+        }
+    }
+
+    /**
+     * A shell wrapper class easy for test.
+     */
+    public class NvidiaCommandExecutor {
+
+        public String getDeviceInfo() throws IOException {
+            return Shell.execCommand(environment,
+                    new String[]{pathOfGpuBinary, "--query-gpu=index,pci.bus_id,mig.mode.current",
+                            "--format=csv,noheader"}, MAX_EXEC_TIMEOUT_MS);
+        }
+
+        public String getDeviceMigInfo() throws IOException {
+            return Shell.execCommand(environment,
+                    new String[]{pathOfGpuBinary, "-L"}, MAX_EXEC_TIMEOUT_MS);
+        }
+
+        public String getMajorMinorInfo(String devName) throws IOException {
+            // output "major:minor" in hex
+            Shell.ShellCommandExecutor shexec = new Shell.ShellCommandExecutor(
+                    new String[]{"stat", "-c", "%t:%T", "/dev/" + devName});
+            shexec.execute();
+            return shexec.getOutput();
+        }
+
+        public void searchBinary() throws Exception {
+            if (pathOfGpuBinary != null) {
+                LOG.info("Skip searching, the nvidia gpu binary is already set: "
```

Comment:
```suggestion
                LOG.info("Skip searching, the NVIDIA GPU binary is already set: "
```

Also, is this supposed to be INFO instead of DEBUG?